### PR TITLE
[SPARK-48675][SQL] Fix cache table with collated column

### PR DESCRIPTION
**Pull request by nikolamand-db**
### What changes were proposed in this pull request?

Following sequence of queries produces the error:
```
> cache lazy table t as select col from values ('a' collate utf8_lcase) as (col);
> select col from t;
org.apache.spark.SparkException: not support type: org.apache.spark.sql.types.StringType@1.
        at org.apache.spark.sql.errors.QueryExecutionErrors$.notSupportTypeError(QueryExecutionErrors.scala:1069)
        at org.apache.spark.sql.execution.columnar.ColumnBuilder$.apply(ColumnBuilder.scala:200)
        at org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer$$anon$1.$anonfun$next$1(InMemoryRelation.scala:85)
        at scala.collection.immutable.List.map(List.scala:247)
        at scala.collection.immutable.List.map(List.scala:79)
        at org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer$$anon$1.next(InMemoryRelation.scala:84)
        at org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer$$anon$1.next(InMemoryRelation.scala:82)
        at org.apache.spark.sql.execution.columnar.CachedRDDBuilder$$anon$2.next(InMemoryRelation.scala:296)
        at org.apache.spark.sql.execution.columnar.CachedRDDBuilder$$anon$2.next(InMemoryRelation.scala:293)
...
```
This is also the problem on non-lazy cached tables.

It turns out that the problem happens to occur during the execution of `InMemoryTableScanExec` where we need to update `ColumnAccessor`, `ColumnBuilder`, `ColumnType` and `ColumnStats`.

### Why are the changes needed?

To fix the described error.

### Does this PR introduce _any_ user-facing change?

Yes, the described sequence of queries should produce valid results after these changes are applied instead of throwing error.

### How was this patch tested?

Added checks to columnar suites for the mentioned classes and integration test to `CollationSuite`.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala [hunk 1]
```diff
@@ -255,23 +256,27 @@ private[columnar] final class DoubleColumnStats extends ColumnStats {
...
       gatherNullStats()
     }
   }
 
   def gatherValueStats(value: UTF8String, size: Int): Unit = {
-    if (upper == null || value.binaryCompare(upper) > 0) upper = value.clone()
-    if (lower == null || value.binaryCompare(lower) < 0) lower = value.clone()
+    def collatedCompare(l: UTF8String, r: UTF8String): Int =
+      CollationFactory.fetchCollation(collationId).comparator.compare(l, r)
+    if (upper == null || collatedCompare(value, upper) > 0) upper = value.clone()
```
Comment (dbatomic) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala:278]: You already have `semanticCompare` in `UTF8String` that does the same thing as your `collatedCompare`.

**Review by dbatomic — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnAccessor.scala [hunk 2]
```diff
@@ -100,8 +100,8 @@ private[columnar] class FloatColumnAccessor(buffer: ByteBuffer)
 private[columnar] class DoubleColumnAccessor(buffer: ByteBuffer)
   extends NativeColumnAccessor(buffer, DOUBLE)
 
-private[columnar] class StringColumnAccessor(buffer: ByteBuffer)
-  extends NativeColumnAccessor(buffer, STRING)
+private[columnar] class StringColumnAccessor(buffer: ByteBuffer, dataType: StringType)
```
Comment (uros-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnAccessor.scala:103]:
```suggestion
private[columnar] class StringColumnAccessor(buffer: ByteBuffer, st: StringType)
```

or

```suggestion
private[columnar] class StringColumnAccessor(buffer: ByteBuffer, stringType: StringType)
```

**Review by uros-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnAccessor.scala (hunk 2 above)
Comment (uros-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnAccessor.scala:103]: either change everywhere or keep as is, I'm fine with either

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/NullableColumnBuilderSuite.scala [hunk 3]
```diff
@@ -39,21 +40,33 @@ object TestNullableColumnBuilder {
 class NullableColumnBuilderSuite extends SparkFunSuite {
   import org.apache.spark.sql.execution.columnar.ColumnarTestUtils._
 
-  Seq(
+  val stringTypes = Seq(
+    STRING(StringType),
```
[the remark this answers is not in the document]
Comment (nikolamand-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/NullableColumnBuilderSuite.scala]: Added comments to several places to make readability better.

**Pull request closed without merging**